src/Workspaces/Core/Portable/CodeActions/CodeAction_Cleanup.cs

```diff
@@ -0,0 +1,180 @@
+        CancellationToken cancellationToken)
+    {
+        // One item per document per cleanup pass.
+        progress.AddItems(documentIdsAndOptions.Length * s_cleanupPasses.Length);
+
+        var currentSolution = solution;
+        foreach (var cleanupPass in s_cleanupPasses)
+            currentSolution = await RunParallelCleanupPassAsync(currentSolution, cleanupPass).ConfigureAwait(false);
+
+        return currentSolution;
+
+        async Task<Solution> RunParallelCleanupPassAsync(
+            Solution solution, Func<Document, CodeCleanupOptions, CancellationToken, Task<Document>> cleanupDocumentAsync)
+        {
+            // We're about to making a ton of calls to this new solution, including expensive oop calls to get up to
+            // date compilations, skeletons and SG docs.  Create and pin this solution so that all remote calls operate
+            // on the same fork and do not cause the forked solution to be created and dropped repeatedly.
+            using var _ = await RemoteKeepAliveSession.CreateAsync(solution, cancellationToken).ConfigureAwait(false);
+
+            return await ProducerConsumer<(DocumentId documentId, SyntaxNode newRoot)>.RunParallelAsync(
+                source: documentIdsAndOptions,
+                produceItems: static async (documentIdAndOptions, callback, args, cancellationToken) =>
+                {
+                    // As we finish each document, update our progress.
+                    using var _ = args.progress.ItemCompletedScope();
+
+                    var (documentId, options) = documentIdAndOptions;
+
+                    // Fetch the current state of the document from this fork of the solution.
+                    var document = args.solution.GetRequiredDocument(documentId);
+                    Contract.ThrowIfFalse(document.SupportsSyntaxTree, "GetDocumentIdsAndOptionsAsync should only be returning documents that support syntax");
+
+                    // Now, perform the requested cleanup pass on it.
+                    var cleanedDocument = await args.cleanupDocumentAsync(document, options, cancellationToken).ConfigureAwait(false);
+                    if (cleanedDocument is null || cleanedDocument == document)
+                        return;
+
+                    // Now get the cleaned root and pass it back to the consumer.
+                    var newRoot = await cleanedDocument.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    callback((documentId, newRoot));
+                },
+                consumeItems: static async (stream, args, cancellationToken) =>
+                {
+                    // Grab all the cleaned roots and produce the new solution snapshot from that.
+                    var currentSolution = args.solution;
+                    await foreach (var (documentId, newRoot) in stream)
+                        currentSolution = currentSolution.WithDocumentSyntaxRoot(documentId, newRoot);
```

Comment:
>WithDocumentSyntaxRoot [](http://example.com/codeflow?start=58&length=22)

FYI, I've got it in my mind that we should eventually be batching these solution updates. I wonder if all these forks are still visible in the profiles with all the changes that have happened since I last measured.

Reply:
yes.  we can totally do that.  it's orthogonal to the work here, but i'm fully in support of it.  

primarily, these prs ensure that we're not forking the solution N times and also getting compilatinos for each fork.

i will do this in a followup  :)